src/Lean/Meta/Tactic/SplitIf.lean

```diff
@@ -56,23 +56,25 @@ def discharge? (useDecide := false) : Simp.Discharge := fun prop => do
          else
            return none
 
-/-- Return the condition of an `if` expression to case split. -/
-partial def findIfToSplit? (e : Expr) : Option Expr :=
+/-- Return the condition (and decidability instance when available) of an `if` expression to case split. -/
+partial def findIfToSplit? (e : Expr) : Option (Expr × Option Expr) :=
   if let some iteApp := e.find? fun e => (e.isIte || e.isDIte) && !(e.getArg! 1 5).hasLooseBVars then
     let cond := iteApp.getArg! 1 5
+    -- If the decidability instance has loose bound variables then we fall back on using classical axioms
+    let inst := if (e.getArg! 2 5).hasLooseBVars then none else some <| iteApp.getArg! 2 5
     -- Try to find a nested `if` in `cond`
-    findIfToSplit? cond |>.getD cond
+    findIfToSplit? cond |>.getD (cond, inst)
   else
     none
 
 def splitIfAt? (mvarId : MVarId) (e : Expr) (hName? : Option Name) : MetaM (Option (ByCasesSubgoal × ByCasesSubgoal)) := do
   let e ← instantiateMVars e
-  if let some cond := findIfToSplit? e then
+  if let some (cond, inst) := findIfToSplit? e then
```

Comment:
```suggestion
  if let some (cond, dec) := findIfToSplit? e then
```
is perhaps a clearer name? Or even `decInst`.

Reply:
I went for `dec?` which also signals it's an `Option Expr` and not an `Expr`.